Fix delivery shipment rate error for unimplemented delivery types

mjrk opened the pull request:
Description of the issue/feature this PR addresses: Unimplemented delivery_types for delivery carriers lead to 500

Current behavior before PR: During an odoo version update or by replacing one delivery app by another, it can happen that the shipping method is not implemented (see delivery_carrier.py#rate_shipment method). This will lead to a 500 Internal Server Error when opening the checkout page that contains the delivery method select.

Desired behavior after PR is merged: Instead of a 500 Internal Server Error (full show stopper), the error message "Shipment rate not found" is returned. The interface displays the error messages like other errors and it is ensured, that the user cannot proceed when selecting an erroneous delivery carrier.

--
I confirm I have signed the CLA and read the PR guidelines at [www.odoo.com/submit-pr](http://www.odoo.com/submit-pr)

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/169008.png)](https://mergebot.odoo.com/odoo/odoo/pull/169008)

vchu-odoo commented:
Hello @mjrk,
Thank you for contributing to Odoo!

I could not reproduce the issue, could you please specify the broken scenario and the delivery carriers that cause this error?

mjrk commented:
@vchu-odoo 

Hello, 

in models/delivery_carrier.py#rate_shipment

    if hasattr(self, '%s_rate_shipment' % self.delivery_type):
        ....

Please see this line of code and see that there is no "else". This effect all and any purchased or included delivery carrier module. 

Scenario: 
We install any delivery carrier with its "delivery_type", configure different delivery methods and use them. Later, we switch to another delivery carrier app. This happens quite often during migrations, if the currently in use delivery carrier app is not available in the new version. But it will also happen in between, when we uninstall the a delivery app. You can use a deliver app of your choice.

If the before-in-use delivery app is uninstalled, its method, for example "dhl_rate_shipment", will not exist anyemore. The if-statement above will return "False". As there is no "else", the method returns "None". If this method returns "None" (instead of a dict with "success" or "error_message"), latter lines, for example "return self.rate_shipment(order).get('success')" will raise an exception, because "None" does not have a "get"-method.

This exception leads to a "Internal Server Error"-page for the clients during checkout in odoo!

Now, in a lot of cases, not programmers, but the clients are responsible to administer their delivery methods. And they may archive or unpublish their previously-in-use delivery methods a little bit later as they should. 

Also note, previously-in-use delivery methods can't be deleted, because of foreign keys. So, If they accidentally unarchive or publish the wrong delivery method - the issue raises again! Then nobody can buy anything in the shop because of an "Internal Server Error" on the checkout page.


Also check the documentation string for the "rate_shipment":

    def rate_shipment(self, order):
        ''' Compute the price of the order shipment

        :param order: record of sale.order
        :return dict: {'success': boolean,
                       'price': a float,
                       'error_message': a string containing an error message,
                       'warning_message': a string containing a warning message}

It is supposed to return a dict. But it will return "None" for a delivery_type if the corresponding delivery app was uninstalled! 


Do you need more information?

vchu-odoo reviewed (commented):
Thank you for your explanation and fix!
There is a valid point and a valid fix, the function should return dict.

Could you please split the commit in two: one for the fix and the other one for adding CLA with [CLA] format in the commit message?

mjrk commented:
@vchu-odoo 

Thank you for your fast response. I split the commit in two, I hope the order (first CLA then FIX) is correct.

vchu-odoo reviewed (approved):
Thank you for the change!

LGTM :+1:

robodoo commented:
Merge method set to rebase and fast-forward.

vava-odoo commented on addons/delivery/models/delivery_carrier.py:
```diff
@@ -240,6 +240,10 @@ def rate_shipment(self, order):
             return res
+        else:
+            return {
+                'error_message': _("Shipment rate not found")
```
Don't we need to export pot file then? Or, in this case, maybe just remove the `_()`?

vava-odoo commented on doc/cla/individual/mjrk.md:
```diff
@@ -0,0 +1,11 @@
+Bulgaria, 27 September, 2021
```
Why not today?

vava-odoo reviewed (commented):
Don't know if important, but wanted to double check with you those two points @AntoineVDV @vchu-odoo 
Also the commit titles do not follow the guidelines :eyes:

AntoineVDV commented:
Thanks @vava-odoo, we staged indeed a bit too fast.

@robodoo r-

@mjrk Could you please address the two issues reported above, and rename your commit titles to "[CLA] signature for mjrk" and "[FIX] delivery: return a shipment rate error message for unimplemented delivery types"? Thanks :pray:

mjrk commented:
@AntoineVDV 

Thank you for your feedback. I changed the date (as requested in https://github.com/odoo/odoo/pull/77288, this PR was rebased from there) and updated the commit messages.

Regarding "export pot file" - I'd keep the possibility to add translation, but I do not understand "export pot file" - does this mean, as part of the commit? Searching for odoo commit guidelines "pot" did not enlighten me until now. 

I can add translations for Bulgarian, German and Spanisch, if this helps.

However, the bug is a bit technical, and removing the _ is totally fine for me, too, if this is preferred.

mjrk commented:
I updated the returned error dict, so it confirms to other returns in Odoo 17. I used the same beginning for the error messages as in other occurrences, because this may facilitate translations.

Please apologize that I did not do it in first place - I just rebased the old PR and tested, that the issue still existed and the fix workes, without doing a full review of changes in the delivery app.

AntoineVDV commented:
When working in stable versions, it is necessary to export the translatable terms of a module to the corresponding .pot​ file when you update those terms. To do so,

1. Start a server with the module installed on the database. (or use [this database](https://63915413-17-0-all.runbot146.odoo.com/) that was created based on this PR)
2. Switch to debug mode.
3. Go to Settings > Translations > Export Translation.
4. In the wizard, select the module in the Apps to Export.
5. Hit Export and download the generated file.
6. Replace the content of the .pot​ file in the sources with that of the downloaded file. Maybe keep only the changes related to this PR, if there is more diff than expected.

Then commit the changes, and it should be good :)

mjrk commented:
@AntoineVDV Please apologize the delay, too much came up, then it got out of focus. Thanks for your reply, I added the pot file.

AntoineVDV reviewed (approved):
Thanks :)

Could you just squash the two last commits (delivery-related) together? Then we're good to go.

mjrk commented:
@AntoineVDV 

Is there anything missing here? I squashed the commits as requested but did not read an update. Now I see the error

> psycopg2.errors.InvalidSavepointSpecification: savepoint "8517e9b2-378a-11ef-a209-0242ac110005" does not exist

Presumably, this is not related to me, squashing commits like requested. I do not find how to re-trigger the check. How do I do this? Or do I have to force push again?

AntoineVDV commented:
Indeed, the branch was likely too old, and some false positives were reported, but I started a new batch of tests, and it's good now.

@robodoo r+ rebase-ff
